db/db_impl/db_impl_compaction_flush.cc

```diff
@@ -2256,19 +2269,36 @@ ColumnFamilyData* DBImpl::PickCompactionFromQueue(
 
 void DBImpl::SchedulePendingFlush(const FlushRequest& flush_req,
                                   FlushReason flush_reason) {
+  mutex_.AssertHeld();
   if (flush_req.empty()) {
     return;
   }
-  for (auto& iter : flush_req) {
-    ColumnFamilyData* cfd = iter.first;
-    cfd->Ref();
-    cfd->SetFlushReason(flush_reason);
+  if (!immutable_db_options_.atomic_flush) {
+    // For the non-atomic flush case, we never schedule multiple column
+    // families in the same flush request.
+    assert(flush_req.size() == 1);
+    ColumnFamilyData* cfd = flush_req[0].first;
+    assert(cfd);
+    if (!cfd->queued_for_flush() && cfd->imm()->IsFlushPending()) {
+      cfd->Ref();
+      cfd->set_queued_for_flush(true);
```

Comment:
Is there any harm in setting/checking this even for atomic flush? Would simplify the code a bit.

Reply:
`queue_for_flush` is not used by atomic flush. In atomic flush, a column family can appear in multiple flush requests in the flush queue at the same time. Using a boolean is not useful because we want to clear this variable when a flush is about to be performed for this column family. There is not an easy way to determine when to clear it, unless we use an integer count (which can be done later in a separate PR). If we set the flag but not clear it, then the assertion for `queued_for_flush_` being false will fail in `~ColumnFamilyData()`.